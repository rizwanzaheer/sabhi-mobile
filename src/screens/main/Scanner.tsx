/**
 * Serto Mobile App
 *
 */
import React, { useState, useContext } from 'react'
import { WalletConnectContext } from '../../providers/WalletConnect'
import { Container, FabButton, Screen } from '@kancha/kancha-ui'
import { RNCamera } from 'react-native-camera'
import { Colors, Icons } from '../../theme'
import { TextInput } from 'react-native-gesture-handler'
import { NEW_MESSAGE } from '../../lib/graphql/queries'
import { useMutation } from '@apollo/react-hooks'

export default (props: any) => {
  const [scannerActive, toggleScanner] = useState(true)
  const [inputMode, toggleInputMode] = useState(false)
  const { walletConnectOnSessionRequest } = useContext(WalletConnectContext)

  const [handleMessage] = useMutation(NEW_MESSAGE, {
    onCompleted(resp) {
      console.log('Success', resp)
    },
    onError(err) {
      if (err) {
        console.log('Error', err)
      }
    },
  })

  const onPasteJWT = (text: string) => {
    onBarCodeRead({ data: text })
  }

  const processDafMessage = async (message: string) => {
    const decodedMessage = await handleMessage({
      variables: {
        raw: message,
        metaData: [{ type: 'qrCode' }],
        save: false,
      },
    })
    const { type, credentials } = decodedMessage.data.handleMessage
    switch (type) {
      case 'w3c.vc':
        props.navigation.dismiss()

        setTimeout(() => {
          props.navigation.navigate('CredentialView', {
            message,
            credentials,
            handleMessage,
          })
        }, 300)

        return
      default:
        props.navigation.navigate('MessageProcess', {
          message,
        })
    }
  }

  const onBarCodeRead = (e: any) => {
    if (scannerActive) {
      if (e.data.startsWith('wc:')) {
        walletConnectOnSessionRequest(e.data)
        props.navigation.dismiss()
      } else {
        processDafMessage(e.data)
      }
    }

    toggleScanner(false)
  }

  return (
    <Screen
      fabButton={
        <Container marginBottom={-16} flexDirection={'row'}>
          <Container flex={1}></Container>
          <Container flex={1} alignItems={'center'}>
            <FabButton
              testID={'CANCEL_SCAN_BTN'}
              color={Colors.CHARCOAL}
              onPress={() => props.navigation.dismiss()}
              icon={Icons.CLOSE}
            />
          </Container>
          <Container flex={1} justifyContent={'flex-start'}>
            {__DEV__ && (
              <FabButton
                testID={'ENABLE_PASTE'}
                color={inputMode ? Colors.CONFIRM : Colors.CHARCOAL}
                onPress={() => toggleInputMode(!inputMode)}
                size={50}
                icon={{ name: 'ios-code', iconFamily: 'Ionicons' }}
              />
            )}
          </Container>
        </Container>
      }
    >
      <Container flex={1} backgroundColor={Colors.BLACK}>
        {inputMode ? (
          <Container justifyContent={'center'} flex={1} padding>
            <TextInput
              testID={'JWT_INPUT'}
              autoCapitalize={'none'}
              autoCorrect={false}
              autoCompleteType={'off'}
              placeholder={'Paste JWT'}
              placeholderTextColor={Colors.MEDIUM_GREY}
              onChangeText={onPasteJWT}
              style={{
                padding: 10,
                color: Colors.LIGHT_GREY,
                backgroundColor: Colors.DARK_GREY,
                borderRadius: 5,
                borderWidth: 1,
              }}
            />
          </Container>
        ) : (
          <RNCamera
            testID={'CAMERA'}
            captureAudio={false}
            style={{ flex: 1 }}
            onBarCodeRead={onBarCodeRead}
          />
        )}
      </Container>
    </Screen>
  )
}
