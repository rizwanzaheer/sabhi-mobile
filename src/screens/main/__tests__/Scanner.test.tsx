import 'react-native'
import React from 'react'
import Scanner from '../Scanner'
import { render, fireEvent, act } from 'react-native-testing-library'
import { MockedProvider } from '@apollo/react-testing'

const navigation = {
  goBack: jest.fn(),
  dismiss: jest.fn(),
  navigate: jest.fn(),
}

jest.useFakeTimers()
jest.runAllTimers()

describe('Scanner', () => {
  it('renders correctly', () => {
    //@ts-ignore
    const tree = render(
      <MockedProvider addTypename={false}>
        <Scanner navigation={navigation} />
      </MockedProvider>,
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('cancel button test', () => {
    //@ts-ignore
    const { getByTestId } = render(
      <MockedProvider addTypename={false}>
        <Scanner navigation={navigation} />
      </MockedProvider>,
    )
    expect(getByTestId('CANCEL_SCAN_BTN')).toBeDefined()

    act(() => {
      fireEvent.press(getByTestId('CANCEL_SCAN_BTN'))
    })

    expect(navigation.dismiss).toBeCalled()
  })

  it('jwt paste handler', () => {
    const component = render(
      <MockedProvider addTypename={false}>
        <Scanner navigation={navigation} />
      </MockedProvider>,
    )
    expect(component.getByTestId('ENABLE_PASTE')).toBeDefined()
    //expect(component.getByTestId('JWT_INPUT')).toThrowError()
    act(() => {
      fireEvent(component.getByTestId('ENABLE_PASTE'), 'press')
    })
    expect(component.getByTestId('JWT_INPUT')).toBeDefined()
  })
})
