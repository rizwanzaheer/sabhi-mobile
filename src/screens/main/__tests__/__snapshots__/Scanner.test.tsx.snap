// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scanner renders correctly 1`] = `
Array [
  <View
    pointerEvents="auto"
    style={
      Object {
        "alignItems": undefined,
        "backgroundColor": undefined,
        "borderRadius": undefined,
        "flex": 1,
        "flexDirection": undefined,
        "height": undefined,
        "justifyContent": undefined,
        "margin": undefined,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "padding": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "width": undefined,
      }
    }
  >
    <View
      pointerEvents="auto"
      style={
        Object {
          "alignItems": undefined,
          "backgroundColor": "#000000",
          "borderRadius": undefined,
          "flex": 1,
          "flexDirection": undefined,
          "height": undefined,
          "justifyContent": undefined,
          "margin": undefined,
          "marginBottom": undefined,
          "marginLeft": undefined,
          "marginRight": undefined,
          "marginTop": undefined,
          "padding": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": undefined,
          "width": undefined,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <ActivityIndicator
          animating={true}
          color="#999999"
          hidesWhenStopped={true}
          size="small"
        />
      </View>
    </View>
  </View>,
  <View
    pointerEvents="auto"
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": undefined,
        "borderRadius": undefined,
        "bottom": 40,
        "flex": undefined,
        "flexDirection": undefined,
        "height": undefined,
        "justifyContent": undefined,
        "margin": undefined,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "padding": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <View
      pointerEvents="auto"
      style={
        Object {
          "alignItems": undefined,
          "backgroundColor": undefined,
          "borderRadius": undefined,
          "flex": undefined,
          "flexDirection": "row",
          "height": undefined,
          "justifyContent": undefined,
          "margin": undefined,
          "marginBottom": -16,
          "marginLeft": undefined,
          "marginRight": undefined,
          "marginTop": undefined,
          "padding": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": undefined,
          "width": undefined,
        }
      }
    >
      <View
        pointerEvents="auto"
        style={
          Object {
            "alignItems": undefined,
            "backgroundColor": undefined,
            "borderRadius": undefined,
            "flex": 1,
            "flexDirection": undefined,
            "height": undefined,
            "justifyContent": undefined,
            "margin": undefined,
            "marginBottom": undefined,
            "marginLeft": undefined,
            "marginRight": undefined,
            "marginTop": undefined,
            "padding": undefined,
            "paddingBottom": undefined,
            "paddingHorizontal": undefined,
            "paddingLeft": undefined,
            "paddingRight": undefined,
            "paddingTop": undefined,
            "width": undefined,
          }
        }
      />
      <View
        pointerEvents="auto"
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": undefined,
            "borderRadius": undefined,
            "flex": 1,
            "flexDirection": undefined,
            "height": undefined,
            "justifyContent": undefined,
            "margin": undefined,
            "marginBottom": undefined,
            "marginLeft": undefined,
            "marginRight": undefined,
            "marginTop": undefined,
            "padding": undefined,
            "paddingBottom": undefined,
            "paddingHorizontal": undefined,
            "paddingLeft": undefined,
            "paddingRight": undefined,
            "paddingTop": undefined,
            "width": undefined,
          }
        }
      >
        <View
          accessibilityLabel="CANCEL_SCAN_BTN"
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#333333",
              "borderRadius": 50,
              "elevation": 3,
              "height": 70,
              "justifyContent": "center",
              "opacity": 1,
              "shadowColor": "#000000",
              "shadowOpacity": 0.3,
              "shadowRadius": 20,
              "width": 70,
            }
          }
          testID="CANCEL_SCAN_BTN"
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#FFFFFF",
                  "fontSize": 35,
                },
                undefined,
                Object {
                  "fontFamily": "Ionicons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
        </View>
      </View>
      <View
        pointerEvents="auto"
        style={
          Object {
            "alignItems": undefined,
            "backgroundColor": undefined,
            "borderRadius": undefined,
            "flex": 1,
            "flexDirection": undefined,
            "height": undefined,
            "justifyContent": "flex-start",
            "margin": undefined,
            "marginBottom": undefined,
            "marginLeft": undefined,
            "marginRight": undefined,
            "marginTop": undefined,
            "padding": undefined,
            "paddingBottom": undefined,
            "paddingHorizontal": undefined,
            "paddingLeft": undefined,
            "paddingRight": undefined,
            "paddingTop": undefined,
            "width": undefined,
          }
        }
      >
        <View
          accessibilityLabel="ENABLE_PASTE"
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#333333",
              "borderRadius": 50,
              "elevation": 3,
              "height": 50,
              "justifyContent": "center",
              "opacity": 1,
              "shadowColor": "#000000",
              "shadowOpacity": 0.3,
              "shadowRadius": 20,
              "width": 50,
            }
          }
          testID="ENABLE_PASTE"
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#FFFFFF",
                  "fontSize": 35,
                },
                undefined,
                Object {
                  "fontFamily": "Ionicons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
        </View>
      </View>
    </View>
  </View>,
]
`;
